Display time as months and days instead of time into season

ehughsbaird opened the pull request:
#### Summary
Features "Show time in months/days"

Built on https://github.com/CleverRaven/Cataclysm-DDA/pull/81492

#### Purpose of change
I think it's more immersive to display the time in months/days.

#### Describe the solution
Because years after the cataclysm are 364 days, and I don't want to do gross calendar math, the structure of the months/days have changed a little:

Seasons remain the same, and the first day of each season is on the 21th of the month. The first full month of a season has 31 days, all other months have 30.

| Month     | Day   | Season |
|-----------|-------|--------|
| January   | 1-31  | Winter |
| February  | 1-30  | Winter |
| March     | 1-20  | Winter |
| March     | 21-30 | Spring |
| April     | 1-31  | Spring |
| May       | 1-30  | Spring |
| June      | 1-20  | Spring |
| June      | 21-30 | Summer |
| July      | 1-31  | Summer |
| August    | 1-30  | Summer |
| September | 1-20  | Summer |
| September | 21-30 | Autumn |
| October   | 1-31  | Autumn |
| November  | 1-30  | Autumn |
| December  | 1-20  | Autumn |
| December  | 21-30 | Winter |

Because calendar::turn_zero is the first day of spring (March 21), some code was needed to add an appropriate offset to ensure the year transitions on Dec 30 -> Jan 1.

Settings for season length other than 91 will not show months.

#### Describe alternatives you've considered
December has 34 days, all other months have 30.

Using the long forms of the month names.

Changing it to display year 0 instead of year 1 (it's 0 years since the cataclysm).

This makes it so you need to know the dates seasons change on to know when they are, and I'd be happy to take suggestions on how to make season change times more visible.

~~I'm very upset October 31 does not exist and I am strongly considering making the first full month of each season have 31 days to fix it.~~

#### Testing

See tests.

#### Additional context
![Image of sidebar showing a game that has just started. The date field says Thursday, May 19](https://github.com/user-attachments/assets/c9d1f9ad-d1b8-4500-bdf1-c1b40ff828f5)
![Image of the date section on the sidebar. It says Sunday, Jun 20](https://github.com/user-attachments/assets/09300cbd-c00d-4957-a9b5-f7e4e48c8d1c)
![Image of the time stamp of a diary entry. It says Year 1, Jun 20, and the time](https://github.com/user-attachments/assets/d6ee4106-d685-43eb-909e-d88281ef8d77)
![Image of the time stamp of a diary entry. It says Year 1, Dec 22](https://github.com/user-attachments/assets/df09c3d9-1702-45b9-8d4d-3240e53023b4)
![Image of the time stamp of a diary entry. It says Year 2, Jan 1, 10 days since last entry.](https://github.com/user-attachments/assets/4eafb10f-1e73-4e9e-98ce-ef4379587690)
![Image of the pages section of a diary. It has entries on Year 1, Jun 20, Year 1, Dec 22, and Year 2, Jan 1](https://github.com/user-attachments/assets/03e4381f-08a3-4b7f-857b-8d39c9e1cec3)

ehughsbaird committed:
Display time with months instead of seasons

Split the 364 day year up into months as so:
The seasons are
Dec, Jan, Feb, Mar -> Winter
Mar, Apr, May, Jun -> Spring
Jun, Jul, Aug, Sep -> Summer
Sep, Oct, Nov, Dec -> Autumn

The first month of each season has 31 days, and the first day of each
season is on the 20th of the first month (the last day on the 19th of
the last month).

For season settings other than 364 days, season display is broken and I
need to consider how to implement this correctly.

Some additional logic was needed for the year transition correctly on
Dec 31 -> Jan 1, because turn_zero is the first day of Spring.

The calendar code isn't too gross, but is still gross.

ehughsbaird committed:
Add test to ensure months/days work properly

Just go through ten years, ensure days go forwards in the right
direction, and months/days transition correctly.

Also add a test to check specific time points look correct.

GuardianDll commented:
#81245
```
#### Describe alternatives you've considered
Tempting to just remove them, but nah.
```
Removing it altogether if it's complicates another code is a valid option fyi

PatrikLundell commented:
I'm not keen on this. It hides the info relevant to game play (season) behind a month obfuscation layer.

I would not recommend this type of change without a rather substantial set of concurrent changes:
- Year of the proper length
- Months of the proper length
- Plant growth cycles (and probably animals too) tied to dates or meteorological conditions, rather than the very coarse seasons.
- Feedback to players and access to info about when plants can be planted (Widely available tables of dates/conditions in "books" available wherever seeds are sold, as well as in homes).

It doesn't exactly help that the game's seasons are out of whack with my (non US) experience of when seasons start and end.

ehughsbaird commented:
> I'm not keen on this. It hides the info relevant to game play (season) behind a month obfuscation layer.

I agree that season should be less hidden and want to make it more obvious. I would appreciate thoughts on how to do this.
 
> I would not recommend this type of change without a rather substantial set of concurrent changes:
> 
>     * Year of the proper length
> 
>     * Months of the proper length

I will not do this and I do not think it is worth the ugly calendar code to make this happen. The existing approach is relatively simple and I think this is a big benefit, and a downside of my changes are that it makes it less simple.

The gameplay effects of the calendar are more important than the display, and the simple 91-day season, 364 day year is easier to program and far less likely to produce bugs than a 365.24 day year.
 
>     * Plant growth cycles (and probably animals too) tied to dates or meteorological conditions, rather than the very coarse seasons.

I think these are good changes, but I don't see how they make it any less obfuscated - I think they would make it significantly more because they are now not all in phase, with a 91-day window.

Most tilesets make the season fairly visible, and without tilesets the color of the tiles of harvestable flora changes.

I am not sure how animals interact with seasons beyond having seasons where they produce young.

I think it would be reasonable to display dates for harvestability and similar in the examine windows of these and will make this change if you think it would improve the visibility you are concerned about.
 
>     * Feedback to players and access to info about when plants can be planted (Widely available tables of dates/conditions in "books" available wherever seeds are sold, as well as in homes).

The only restriction on agriculture is that it is more than 10 degrees C outside for planting.

https://github.com/CleverRaven/Cataclysm-DDA/blob/1cff63a5ef011410c4d0046a8aea247c0fbd2d89/src/weather.cpp#L874-L883

PatrikLundell commented:
The calculations are performed in the players' heads. The actual information provided is that the sale inventory will be updated in e.g. 4 days and 15 hours. So, what date should you return when the current date is displayed as Feb 27? Easy, the 4th of March of course (rounding the hours up), except during leap years and when weirdo month length is in play...
You ought to change from reporting a duration to reporting an absolute time (and force those players who want to know how many days that is to try to calculate the number of days from that info, i.e. the reverse calculation).

It's not relevant to a work that's only intended to make the display use weirdo months rather seasons, while retaining and hiding the season logic underneath it.
It is, however, relevant to changes which would actually make months relevant and useful.

ehughsbaird committed:
NPCs say restock in absolute time

So it's more clear when the player should come back to restock.

ehughsbaird committed:
Add an option to show seasons instead of month/day

PatrikLundell has some reasoned objections to the months/day scheme, so
add an option to switch to the season scheme.

ehughsbaird commented:
Okay, I have added an option to change the display to seasons.

Outstanding issues:

- [ ] display season in some way
- [ ] fix october 31

I-am-Erk commented:
I was under the impression there'd been a previous 'no' to this change but I don't recall what the reason was. @kevingranade may remember

RenechCDDA commented:
> I was under the impression there'd been a previous 'no' to this change but I don't recall what the reason was. kevingranade may remember

From the discussions on discord, that rejection was on a strict tracking of the real world calendar. This PR is a highly abstracted version of the calendar which has been specifically configured to avoid all the awfulness of real timekeeping. I don't expect any issues on this.

https://discord.com/channels/598523535169945603/598529174302490644/1252074736461283469

<img width="1806" height="259" alt="image" src="https://github.com/user-attachments/assets/4b2749a5-7ba5-4f7c-8385-493fea80905e" />

submariwastaken1time commented:
> Okay, I have added an option to change the display to seasons, and NPCs say the time of next update in absolute time.
> 
> Outstanding issues:
> 
> * [ ]  display season in some way
> * [x]  fix october 31

what about displaying the season as an icon or text do you not like?

GuardianDll commented:
What's the status of this PR?

ehughsbaird committed:
Add diary summary page

This displays some stats about the time spent playing the game, as well
as the current season and how much longer it will last.

This allows viewing the current season and time left in it when the time
display is set to months.

sparr commented:
Diary summary page makes all the "this is the first page" code paths under get_change_list dead, including putting "X days since last entry" on the first added page, showing initial stats and skills as changes, etc.

Starting discussion on Discord at https://discord.com/channels/598523535169945603/598535827169083403/1404444342487351469

I'll probably fix this stuff in #82301 if we can agree on a path forward.
